go-selinux/pwalk.go

```diff
@@ -0,0 +1,128 @@
+// +build selinux,linux
+
+package selinux
+
+/* The code below was taken from
+ * https://github.com/stretchr/powerwalk/blob/master/walker.go
+ * and modified to use github.com/karrick/godirwalk
+ */
+
+import (
+	"sync"
+
+	"github.com/karrick/godirwalk"
+	"github.com/pkg/errors"
+)
+
+// DefaultConcurrentWalks is the default number of files that will be walked at the
+// same time when the Walk function is called.
+// To use a value other than this one, use the WalkLimit function.
+const DefaultConcurrentWalks int = 100
+
+// Walk walks the file tree rooted at root, calling walkFn for each file or
+// directory in the tree, including root. All errors that arise visiting files
+// and directories are filtered by walkFn. The output is non-deterministic.
+// WalkLimit does not follow symbolic links.
+//
+// For each file and directory encountered, Walk will trigger a new Go routine
+// allowing you to handle each item concurrently.  A maximum of DefaultConcurrentWalks
+// walkFns will be called at any one time.
+func Walk(root string, options *godirwalk.Options) error {
+	return WalkLimit(root, options, DefaultConcurrentWalks)
+}
+
+// WalkLimit walks the file tree rooted at root, calling walkFn for each file or
+// directory in the tree, including root. All errors that arise visiting files
+// and directories are filtered by walkFn. The output is non-deterministic.
+// WalkLimit does not follow symbolic links.
+//
+// For each file and directory encountered, Walk will trigger a new Go routine
+// allowing you to handle each item concurrently.  A maximum of limit walkFns will
+// be called at any one time.
+func WalkLimit(root string, options *godirwalk.Options, limit int) error {
+	// make sure limit is sensible
+	if limit < 1 {
+		panic("powerwalk: limit must be greater than zero.")
```

Comment:
this should probably be an error, not a panic

Reply:
agree